### PR TITLE
move NAT mapping logic out of the host, add tests for NAT handling

### PR DESCRIPTION
**Pull request by marten-seemann**
In preparation for #2229. 

There's a few packages involved in NAT handling here:
1. [go-nat](https://github.com/libp2p/go-nat): Wraps UPnP and PMP libraries and provides functions to add, query and remove NAT mappings.
2. `p2p/net/nat`: Uses `go-nat` periodically renew NAT mappings and query for IP address changes.
3. `p2p/host/basic/natmgr.go`: Listens for `Listen` / `ListenClose` notifications from the swarm and tries to obtain NAT mappings for our listen addresses. Provides a `GetMapping(ma.Multiaddr) ma.Multiaddr` method that applies NAT mappings to existing multiaddresses.

### CODE REVIEW

File: p2p/host/basic/basic_host.go [hunk 1]
```diff
@@ -849,99 +848,19 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
...
-			naddr, err := manet.ToNetAddr(transport)
-			if err != nil {
-				log.Error("error parsing net multiaddr %q: %s", transport, err)
+			extMaddr := h.natmgr.GetMapping(listen)
+			if extMaddr == nil {
+				// not mapped
 				continue
 			}
 
-			var (
```
Comment (MarcoPolo) [inline p2p/host/basic/basic_host.go:904]: 🙏 . Thanks for getting rid of this

File: p2p/host/basic/mockgen_private.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# These interfaces might be used as embedded interfaces,
+# so we need to pass them to mockgen as aux_files.
+AUX=()
+for f in *.go; do
+  if [[ -z ${f##*_test.go} ]]; then
+    # skip test files
+    continue;
+  fi
+  if $(egrep -qe "type (.*) interface" $f); then
+    AUX+=("github.com/quic-go/quic-go=$f")
```
Comment (MarcoPolo) [inline p2p/host/basic/mockgen_private.sh:20]: Do we need to change this?

File: p2p/host/basic/mockgen_private.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
+DEST=$2
```
Comment (MarcoPolo) [inline p2p/host/basic/mockgen_private.sh:3]: Did you copy this from quic-go? I'm curious how deeply this needs to be reviewed

**Review by MarcoPolo — COMMENTED** [review]
Couple comments, but looks good. Thanks for the tests!

File: p2p/host/basic/mockgen_private.sh (hunk 3 above)
Comment (marten-seemann) [reply p2p/host/basic/mockgen_private.sh:3]: I did, and it's super ugly. I figured out a better solution though (thanks ChatGPT): We can just export the type in a file that has a `gomock` build flag, and then run `mockgen` with the `-build_flags="-tags=gomock"` command line parameter.
That way, `gomock` will think that it's an exported type, whereas the final build (and, very importantly, your IDE) will not.

Comment (marten-seemann) [reply p2p/host/basic/mockgen_private.sh:3]: In fact, I should do the same in quic-go. This bash script has been annoying me for a long time!

File: p2p/net/nat/nat.go [hunk 4]
```diff
@@ -40,7 +53,20 @@ func DiscoverNAT(ctx context.Context) (*NAT, error) {
...
+		nat:       natInstance,
+		extAddr:   extAddr,
+		mappings:  make(map[entry]int),
+		ctx:       ctx,
+		ctxCancel: cancel,
+	}
+	nat.refCount.Add(1)
+	go func() {
+		defer nat.refCount.Done()
+		nat.background()
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply p2p/net/nat/nat.go:67]: I fixed the double decrement.

File: p2p/host/basic/mockgen_private.sh (hunk 2 above)
Comment (marten-seemann) [reply p2p/host/basic/mockgen_private.sh]: Done! It's so much nicer now!

**Review by MarcoPolo — APPROVED** [review]
Thanks for getting rid of the bash script :)

**Pull request merged**